jerry-core/ecma/base/ecma-helpers-string.c

```diff
@@ -2482,6 +2507,185 @@ ecma_string_trim (const ecma_string_t *string_p) /**< pointer to an ecma string
+ *
+ * @return pointer to the end of the data in the underlying buffer
+ */
+static lit_utf8_byte_t *
+ecma_stringbuilder_grow (ecma_stringbuilder_t *builder_p, /**< string builder */
+                         lit_utf8_size_t required_size) /**< required size */
+{
+  ecma_stringbuilder_header_t *header_p = builder_p->header_p;
+
+  const lit_utf8_size_t new_size = header_p->current_size + required_size;
+  header_p = jmem_heap_realloc_block (header_p, header_p->current_size, new_size);
+  header_p->current_size = new_size;
+  builder_p->header_p = header_p;
+
+  return ((lit_utf8_byte_t *)  header_p) + header_p->current_size - required_size;
+} /* ecma_stringbuilder_grow */
+
+/**
+ * Append an ecma_string_t to a string builder
+ */
+void
+ecma_stringbuilder_append (ecma_stringbuilder_t *builder_p, /**< string builder */
+                           const ecma_string_t *string_p) /**< ecma string */
+{
+  const lit_utf8_size_t string_size = ecma_string_get_size (string_p);
+  lit_utf8_byte_t *dest_p = ecma_stringbuilder_grow (builder_p, string_size);
+
+  size_t copied_size = ecma_string_copy_to_cesu8_buffer (string_p,
+                                                         dest_p,
+                                                         string_size);
+  JERRY_ASSERT (copied_size == string_size);
+} /* ecma_stringbuilder_append */
+
+/**
+ * Append a magic string to a string builder
+ */
+void
+ecma_stringbuilder_append_magic (ecma_stringbuilder_t *builder_p, /**< string builder */
+                                 const lit_magic_string_id_t id) /**< magic string id */
+{
+  const lit_utf8_size_t string_size = lit_get_magic_string_size (id);
+  lit_utf8_byte_t *dest_p = ecma_stringbuilder_grow (builder_p, string_size);
+
+  const lit_utf8_byte_t *string_data_p = lit_get_magic_string_utf8 (id);
+  memcpy (dest_p, string_data_p, string_size);
+} /* ecma_stringbuilder_append_magic */
+
+/**
+ * Append raw string data to a string builder
+ */
+void
+ecma_stringbuilder_append_raw (ecma_stringbuilder_t *builder_p, /**< string builder */
+                               const lit_utf8_byte_t *data_p, /**< pointer to data */
+                               const lit_utf8_size_t data_size) /**< size of the data */
+{
+  lit_utf8_byte_t *dest_p = ecma_stringbuilder_grow (builder_p, data_size);
+  memcpy (dest_p, data_p, data_size);
+} /* ecma_stringbuilder_append_raw */
+
+/**
+ * Append an ecma_char_t to a string builder
+ */
+void
+ecma_stringbuilder_append_char (ecma_stringbuilder_t *builder_p, /**< string builder */
+                                const ecma_char_t c) /**< ecma char */
+{
+  const lit_utf8_size_t size = (lit_utf8_size_t) lit_char_get_utf8_length (c);
+  lit_utf8_byte_t *dest_p = ecma_stringbuilder_grow (builder_p, size);
+
+  lit_char_to_utf8_bytes (dest_p, c);
+} /* ecma_stringbuilder_append_char */
+
+/**
+ * Finalize a string builder
+ *
+ * @return the created string from the underlying buffer
```

Comment:
It would be good to document that after this do not use the `builder_p` as it is released.